rpc/src/client.rs

```diff
@@ -177,52 +201,89 @@ mod wasm {
     use send_wrapper::SendWrapper;
     use serde::Serialize;
     use serde::de::DeserializeOwned;
+    use tracing::warn;
+    use web_sys::AbortController;
 
     use crate::Error;
 
+    const ABORT_ERROR_NAME: &str = "AbortError";
+
     /// Json RPC client.
     pub struct Client {
         id: AtomicU64,
         url: String,
         auth_token: Option<String>,
+        timeout_ms: Option<u32>,
     }
 
+    #[bon]
     impl Client {
         /// Create a new Json RPC client.
         ///
         /// Only the 'http\[s\]' protocols are supported because JavaScript
         /// doesn't allow setting headers with websocket. If you want to
         /// use the websocket client anyway, you can use the one from the
         /// `jsonrpsee` directly, but you need a node with `--rpc.skip-auth`.
-        pub async fn new(conn_str: &str, auth_token: Option<&str>) -> Result<Self, Error> {
-            let protocol = conn_str.split_once(':').map(|(proto, _)| proto);
+        #[builder]
+        pub async fn new(
+            url: &str,
+            auth_token: Option<&str>,
+            connect_timeout: Option<Duration>,
+            request_timeout: Option<Duration>,
+        ) -> Result<Self, Error> {
+            let protocol = url.split_once(':').map(|(proto, _)| proto);
             match protocol {
                 Some("http") | Some("https") => (),
-                _ => return Err(Error::ProtocolNotSupported(conn_str.into())),
+                _ => return Err(Error::ProtocolNotSupported(url.into())),
             };
+            let timeout_ms = request_timeout
+                .map(|t| t.as_millis().try_into())
+                .transpose()
+                .map_err(|_| Error::TimeoutOutOfRange)?;
+            if connect_timeout.is_some() {
+                warn!("ignored connect_timeout for wasm");
+            }
 
             Ok(Client {
                 id: AtomicU64::new(0),
-                url: conn_str.into(),
+                url: url.into(),
                 auth_token: auth_token.map(ToOwned::to_owned),
+                timeout_ms,
             })
         }
 
-        // TODO: Add request timeouts
         async fn send<T: Serialize>(&self, request: T) -> Result<JsResponse, ClientError> {
             let fut = {
                 let mut req = JsRequest::post(&self.url);
 
+                if let Some(timeout) = self.timeout_ms {
+                    let abort_controller = AbortController::new().unwrap(); // FIXME: unwrap
```

Comment:
do you plan on doing the `FIXME` in this scope?

Reply:
AbortController is marked as baseline widely available in mdn, so switched to expect here.